File: compose/sample/src/main/java/br/com/zup/nimbus/todo/app/components/filterNotes.kt
/*
 * Copyright 2023 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.nimbus.todo.app.components

import br.com.zup.nimbus.annotation.AutoDeserialize

@AutoDeserialize
fun filterNotes(
    notes: List<NoteSection>,
    text: String,
    todo: Boolean,
    done: Boolean,
): List<Any> {
    val result = notes.map { section ->
        val filtered = section.items.filter {
            val matchesTextFilter =
                text.isBlank() ||
                it.description.contains(text) ||
                it.title.contains(text)
            val matchesDoneFilter = (todo && !it.isDone) || (done && it.isDone)
            matchesTextFilter && matchesDoneFilter
        }
        NoteSection(section.date, filtered)
    }.filter { it.items.isNotEmpty() }

    return result.map { it.toMap() }
}

File: swift-ui/TodoApp/Operations/FilterNotes.swift
import NimbusSwiftUI

struct Note: Decodable {
  var id: Int
  var title: String
  var description: String
  var date: Int
  var isDone: Bool

  func toMap() -> [String: Any] {
    [
      "id": id,
      "title": title,
      "description": description,
      "date": date,
      "isDone": isDone
    ]
  }
}

struct NoteSection: Decodable {
  var date: Int
  var items: [Note]

  func toMap() -> [String: Any] {
    [
      "date": date,
      "items": items.map { $0.toMap() }
    ]
  }
}

struct SaveNote: OperationDecodable {
  static var properties = ["notes", "noteToSave"]

  var notes: [NoteSection]
  var noteToSave: Note

  func execute() -> [Any] {
    let savedNotes: [NoteSection] = []
    notes.forEach { section in
      let savedSection = NoteSection()
      section.items.forEach { item in
        if (item.id == noteToSave.id)
      }
    }
    
  }
}
